Dyn array: the basics™

c-cube opened the pull request:
:woman_dancing:

dhil committed:
Expose fiber primitives for continuation cloning

This PR makes available the `alloc_stack_noexc` and
`rewrite_exception_stack` from `runtime/fiber.c` in the C API. In
accordance with the C API the functions are renamed to
`caml_alloc_stack_noexc` and `caml_rewrite_exception_stack`,
respectively.

My motivation for this change is to be able to support multi-shot
continuations in OCaml via a library. I am interested in exploring
design and programming with multi-shot continuations in my own
research. To facility this research, I am in the process of building a
[library](https://github.com/dhil/ocaml-multicont) that provides
multi-shot continuations on top of OCaml's linear continuations. Under
the hood, the library uses a variation of the `clone_continuation`
primitive that used to reside in the `Obj` module in previous versions
of Multicore OCaml. In order to implement `clone_continuation` as
library I need access to the two above mentioned fiber manipulation
primitives in `runtime/fiber.c` which are currently marked
`static`. It is not clear to me whether there is a good reason for
having them as `static`, whereas if they are non-static then I am able
to implement continuation cloning externally. I need access to
`alloc_stack_noexc` in order to allocate enough space for a
carbon-copy of an existing stack segment, and
`rewrite_exception_stack` in order to patch up exception pointers for
cloned stack segments.

In my current implementation I duplicate the implementations of these
primitives in order to get going. Obviously, it would be a lot better
and more robust if I could simply reuse the implementations in the
OCaml runtime directly.

As of writing, this PR exposes only the two primitives that I need to
get my minimal implementation of multi-shot continuations
going. However, I think there is a boarder discussion to be had about
whether we should make available the rest of the fiber manipulation
primitives in `runtime/fiber.c` too.

c-cube committed:
first implementation of Dyn_array

c-cube committed:
have dyn array compile, add docs, remove unused functions

c-cube committed:
integrate dyn_array to stdlib's list of modules

c-cube committed:
add some basic tests for Dyn_array

c-cube committed:
tests for ensure

c-cube commented:
It compiles and the tests are actually passing now!

Octachron commented:
Thanks!

I think that the test might be clearer as an expect test? I can do the conversion if we agree.

I will try to have a look at implementing the junk_filling function on the C side soon.

gadmm committed:
Assert Caml_state != NULL

gadmm committed:
Make CAML(un)likely non-internal and use it inside CAMLassert

gadmm committed:
Introduce run-time checks that the domain lock is held in the C API

gadmm committed:
Document `Caml_state_opt`

Implement suggestions from Gabriel Scherer and Damien Doligez

Octachron committed:
Merge pull request #10894 from dhil/expose-fiber-primitives

Expose fiber primitives for continuation cloning

OlivierNicole committed:
Avoid data race in method caching in ocamlrun (#11512)

When calling a method with the opcode GETPUBMET, the method address is
resolved and possibly cached by modifying a word in the bytecode. If
multiple domains execute the same GETPUBMET opcode concurrently, this is
technically a data race. However, any integer value in the correct range
is a safe value. Therefore we inform the compiler of potential
concurrent accesses by using atomic accesses with relaxed memory order.

damiendoligez committed:
fix #11482: random crash in large closure allocation (#11542)

dra27 committed:
Remove utils/config.generated.ml during distclean

Generated by configure, but was still being erroneously deleted during
partialclean.

dra27 committed:
Merge pull request #11549 from dra27/fix-bootstrap

Remove utils/config.generated.ml during distclean

dra27 committed:
Fix configuration of ocamltest without Unix

Missing case in configure.ac

dra27 committed:
quotecommand.ml test doesn't need unix

dra27 committed:
caml-tex tests need unix and str

dra27 committed:
Merge pull request #11551 from dra27/fix-ocamltest-no-unix

Fix ocamltest built without unix

abbysmal committed:
otherlibs/systhreads: domain_id is always int, not uintnat (#11550)

gasche committed:
Domain.recommended_domain_count with type unit -> int (#11545)

Presenting this as a function rather than a constant in the stdlib
leaves us with more implementation options open in the future.

gasche committed:
Merge pull request #11506 from gadmm/caml_state_assertion

Introduce run-time checks that the domain lock is held

dra27 committed:
Update frame-pointers test reference files

Alternate signal stack related changes.

dra27 committed:
Merge pull request #11553 from dra27/frame-pointers-references

Update frame-pointers test reference files

xavierleroy committed:
Move `Is_young` back to `<caml/address_class.h>` and export it

For backward compatibility with OCaml 4.

Fixes: #11464

dra27 committed:
Merge pull request #11555 from xavierleroy/Is_young-is-back

Move `Is_young` back to `<caml/address_class.h>` and export it

FardaleM committed:
documentation: improve the sync_stdlib_docs tool

Try to improve the case when there is function with label argument in a
non label module.

dra27 committed:
Merge pull request #11556 from FardaleM/doc_typo

documentation: fix typos

xavierleroy committed:
caml_make_float_vect in no-flat-float array: use static allocation (#11552)

Allocate the initializing float statically and outside the heap,
rather than dynamically at first use.  Not only it's more efficient,
but it also avoids a race condition.

Fixes: #11504

c-cube committed:
new design without Obj; rename dummy to filler

c-cube committed:
update tests

c-cube committed:
hygiene

c-cube committed:
add @since and comments

c-cube committed:
more doc

c-cube committed:
changes

c-cube committed:
blurb

c-cube committed:
hygiene

c-cube committed:
fix: next_grow_ must actually grow for len=1

c-cube committed:
add test to expose former bad behavior on `next_grow_`

c-cube committed:
rename Dyn_array to Dynarray

c-cube committed:
fix vertical spacing

c-cube committed:
rename pop_exn/pop to pop_last/pop_last_opt, add remove_last

c-cube committed:
use `push_back` instead of `push`

c-cube committed:
fix tests for dynarray

c-cube committed:
rename push_back to push_last

c-cube committed:
remove uses of `Array.unsafe_{get,set}` in dynarray

c-cube committed:
more docs

c-cube committed:
fix .mli

c-cube committed:
add filter/filter_map to dynarray

c-cube committed:
more docs

Octachron commented:
I am closing this draft since the main PR has been opened.
